proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,861 @@
+Indeed, users are free to define their own alternative linear types
+enabled variant of the `Prelude` and even the rest of `base`. To the
+extent that these alternative preludes just reexport data type
+definitions while changing only the types of functions, these preludes
+will not contribute to splintering the ecosystem any more than the
+myriad existing alternative preludes do.
+
+Formalism
+~~~~~~~~~
+
+So far, we have considered only two multiplicities ``1`` and
+``ω``. But the metatheory works with any so-called
+sup-semi-lattice-ordered semi-ring of multiplicities. That is: there
+is a 0, a 1, a sum and a product with the usual distributivity laws, a
+(computable) order compatible with the sum and product, such that each
+pair of multiplicities has a (computable) join. Even if there is only
+three multiplicities in this proposal, the proposal is structured to
+allow future extensions.
+
+Non of our examples use ``0``, however, ``0`` turns out to be useful
+for implementation of type-checking. Additionally, ``0`` has been used
+by `Conor McBride
+<https://link.springer.com/chapter/10.1007/978-3-319-30936-1_12>`_ to
+handle dependent types, which may matter for Dependent Haskell). In
+both case, the use of ``0`` could be seen as an internal use, but
+there is no real reason to deny access to the programmer. Hence it is
+included in the syntax.
+
+Here is the definition of sum, product and order for this proposal's
+multiplicities (in Haskell pseudo-syntax):
+
+::
+   0 + x = x
+   x + 0 = x
+   _ + _ = ω
+
+   0 * _ = 0
+   _ * 0 = 0
+   1 * x = x
+   x * 1 = 1
+   ω * ω = ω
+
+   _ ⩽ ω = True
+   x ⩽ y = x == y
+
+Every variable in the environment is annotated with its multiplicity,
+which constrains how it can be used. A variable usage is said to be
+multiplicity ``p`` in a term ``u`` if:
+
+- ``p=0`` and ``x`` is not free in ``u``
+- ``p=1`` and ``u = x``
+- ``p=p1+q*p2`` `` and ``u = u1 u2`` with ``u1 :: a :q-> b`` and the
+  usage of ``x`` in ``u1`` is ``p1``, and in ``u2`` is ``p2``
+- ``u = λy. v`` and the usage of ``x`` in ``v`` is ``p``.
+
+A variable's usage is correct if it is smaller than or equal to the
+multiplicity annotation of the variable. Incorrect usage results in a
+type error.
+
+The multiplicity of a variable introduced by a λ-abstraction is taken
+from the surrounding typing information (typically a type annotation
+on an equation). For instance
+
+::
+
+  foo :: A :p-> B
+  foo x = …  -- x has multiplicity p
+
+The above takes care of the pure λ-calculus part of Haskell. We also
+need to consider ``let`` and ``case``.
+
+A ``let`` binding is considered to have an implicit multiplicity
+annotation (the annotation is inferred). The variables introduced by a
+``let`` bindings with annotation ``p`` all have multiplicity
+``p``. And the usage of ``x`` in ``let_p {y1 = u1; … ;yn = un} in v``
+is ``p*q1 + … + p*qn + q`` where the usage of ``x`` in ``ui`` is
+``qi`` and in ``v`` is ``q``.
```

Comment:
Please, clarify what `yi` stands for. I'm guessing it can't be a pattern but only a variable or a function definition.

Reply:
In this case, they are meant to be variables yes. Will edit.

Reply:
Done.